client/pov-recovery/src/lib.rs

```diff
@@ -408,15 +450,9 @@ where
 						return;
 					}
 				},
-				imported = imported_blocks.next() => {
```

Comment:
I dont think this is save to remove. Before we recover a block we check if the block has been imported in the meantime here:
https://github.com/paritytech/cumulus/blob/55520786c37f57600102ecfa6a008fb92add9a0e/client/pov-recovery/src/lib.rs#L233

Without removing the imported blocks from `pending_candidates`, we would always trigger a recovery even if the block is known by then.

Reply:
Good point. Indeed we may trigger the recovery of something that in the meantime has been already imported.

I'm open to suggestion, but since an imported block may unexpectly "disappear" under our foot because of the "limit monitor" (or some other future reason) maybe the safer way is to re-check the state after the delay? Or just set a bool in the `PendingCandidate` to prevent auto recovery after the delay. E.g. a `pending_recovery` bool.

The entry is not removed on "block_import" anymore because we may need it to reconstruct the chain of missing entries in the backend (I added a `parent` entry in the `PendingCandidate` struct)

Reply:
Maybe we need to rethink the semantics of `pending_candidates`. Before this PR it was only adding unknown blocks. 

After your change, it is adding every candidate we see, and only removes on finalization or recovery. So blocks that are already known on notification stay until finalization. 

One idea could be to change `pending_candidates` to contain all pending candidates and remove only on finalization. We would need a separate cache that has all the imported hashes of the last $RecoveryDelay seconds to do the import check. I think this would also solve the problem from the other comment https://github.com/paritytech/cumulus/pull/1891#discussion_r1028517177

Reply:
Addressed by https://github.com/paritytech/cumulus/pull/1891/commits/e7660ea44908a183964f69e19019f1ccb4d250ec